app/code/Magento/InventoryConfigurableProduct/Observer/ProcessSourceItemsObserver.php

```diff
@@ -53,10 +61,17 @@ public function execute(EventObserver $observer)
 
         if ($configurableMatrix != '') {
             $productsData = json_decode($configurableMatrix, true);
-            foreach ($productsData as $productData) {
-                if (isset($productData['quantity_per_source']) && is_array($productData['quantity_per_source'])) {
-                    $sku = $productData[ProductInterface::SKU];
-                    $this->processSourceItems($productData['quantity_per_source'], $sku);
+            foreach ($productsData as $key => $productData) {
+                if (isset($productData['quantity_per_source'])) {
+                    $quantityPerSource = is_array($productData['quantity_per_source'])
+                        ? $productData['quantity_per_source']
+                        : [];
+
+                    // get sku by child id, because child sku can be changed if product with such sku already exists.
+                    $childProductId = $product->getExtensionAttributes()->getConfigurableProductLinks()[$key];
+                    $childProduct = $this->productRepository->getById($childProductId);
```

Comment:
We don't need to load product
We have `\Magento\InventoryCatalog\Model\GetSkusByProductIdsInterface` for resolving sky by product id

Reply:
Fixed.